### PR TITLE
N `v0.0`

### LINKED ISSUE #218: Mutex
Since #166 is not finalized yet, I think it might be easier to implement mutexes first.

[Mutexes](https://en.wikipedia.org/wiki/Lock_(computer_science)) contain a value, but in order to access it, it must be locked, preventing other threads from accessing it. Then, the thread with the lock can read and write the value as it pleases, then it closes the lock so that it can be accessed by other threads again. This makes it safe without encountering scary race conditions and other issues, while also allowing a thread to hold a mutable value.

Because locking might make the function wait until it unlocks, lock functions are cmds.

Also, I think we should put these functions in a native module so that all of the mutex functions are namespaced.

```ts
import mutex

// Create a new conch
assert type mutex.new : [t] t -> mutex.locked[t]

// Locks the conch, preventing other threads from reading from/writing to the conch. If the conch is already locked, it'll wait until it's unlocked. The given function can access the value inside the conch and read or write to it as it pleases, then return a value for the outer function to use.
assert type mutex.access : [a, b] (mutex.unlocked[a] -> cmd[b]) -> mutex.locked[a] -> cmd[b]

// Similar to conch.access, but if the conch is already locked, the cmd will resolve immediately with `none`.
assert type mutex.tryAccess : [a, b] (mutex.unlocked[a] -> cmd[b]) -> mutex.locked[a] -> cmd[maybe[b]]

// If we pursue option 2, then we'll need these functions to read/write the unlocked conch

// Read from the conch
assert type mutex.read : [t] mutex.unlocked[t] -> cmd[t]

// Write a value to the conch.
assert type mutex.write : [t] t -> mutex.unlocked[t] -> cmd[t]
```

### LINKED ISSUE #225: Add features from TypeScript
Things to add (May be changed in the future):
- [x] `none` being falsey
- [x] `?` record access operator
- [x] `??` as alias for `default` (might be changed to `|`)
- [x] ~~Add enums for values\*~~
- [x] ~~Allow for `maybe` typed fields in records to be ommitted~~ (unable to exist with the current system)

\*Example:
```ts
type Test = "a" | "b" | "c"
```

### LINKED ISSUE #238: Proper procedure for printing to stdout
`SystemIO` should have a procedure that prints text or bytes to stdout.

Why not `print`? `print` does not return a `cmd`, so it's an impure function that is used like a pure function only for debugging purposes. In production, the `print` statement will probably be optimised away.

So we should add

```js
import SystemIO

// Write bytes to stdout
assert type SystemIO.write : list[int] -> cmd[()]

// Write text (in UTF-8) to stdout and flushes
assert type SystemIO.print : str -> cmd[()]
```

Ideally, `SystemIO.write`, which takes bytes to write to stdout, would take a byte array, but those aren't in N yet. A special byte array type would be ideal since it would take up less memory (Python ints are bigints, and JS numbers are 64-bit floats rather than 8-bit unsigned integers) as well as being more performant. We also wouldn't have to do strange things to integers outside the 0-255 range. However, `FileIO`'s methods all use `list[int]`, so we might as well be consistent

We should figure out:

- What do we do with a `list[int]` containing values outside of the 0-255 range? Is it undefined behaviour? (There's no test for `FileIO.*Bytes` for this I believe)
- Should `SystemIO.print` add a newline to the end or should we add a separate `SystemIO.println` procedure?
- How do we test these procedures using `assert value`?

### LINKED ISSUE #239: Allow `!` to be used as a statement
`let _ = cmd!` is verbose, and I'm thinking of making a warning about implicitly dropping `cmd` values while suggesting to use `let _ =` to suppress the warning.

```js
Error: This returns a `cmd`, but you didn't do anything with it, so it'll never execute.

3 | times.sleep(1000)
    ^^^^^^^^^^^^^^^^^

Since you're inside a procedure, you can execute the `cmd` using the `!` operator.

If this is intentional, discard the `cmd` explicitly: `let _ = ...`.
```

### LINKED ISSUE #257: Scope object is printed at the end of execution
This is because of the check `if errors is not None`, which causes the scope to print out at the end.

### LINKED ISSUE #259: Breaking change: change function argument syntax
- Use parentheses rather than square brackets
- Require commas between name-type pairs
- Argument and return types can be inferred, so syntactically they're optional

See https://discord.com/channels/710932856251351111/710936824121655396/878023177572347936

The following should have no syntax errors. Note that this is just for a syntax test; these examples may not necessarily pass type checking.

```ts
let square = (n) -> { return n * n }

let add = (a: float, b: float) -> { return a + b }

let getSchedule = ({ schedule }) -> { return schedule }
```

Not in this issue:

- How the argument/return types are inferred
- Allowing type annotations inside patterns (eg `((a: int, b: int)) -> ...`)
- Whether parentheses around a single argument can be omitted
- A shorthand function expression that lets you just give an expression with an implicit return statement

### LINKED ISSUE #260: Allow using `..` on tuples in function expressions
See https://discord.com/channels/710932856251351111/710936824121655396/878023177572347936

```
let printNums = [a:int b:int c:int] -> cmd[()] {
  assert value a == 1
  assert value b == 2
  assert value c == 3
}

let tuple = (1, 2)

let pub main = printNums (..tuple, 3)
```

Not in this issue:
- Using `..` on tuples within tuple expressions

### LINKED ISSUE #263: Allow accessing record fields directly on `maybe` types if they contain a record type
The field will return a `maybe`.

See https://discord.com/channels/710932856251351111/710936824121655396/878032104980615299

```ts
assert type yes({ a: 1 }).a : maybe[int]

assert value yes({ a: 1 }).a == yes(1)

// Should we allow this? The type system would have to be a bit clever here.
assert value none.a == none
```

### LINKED ISSUE #264: Supply default value to a `maybe` using the OR operator
Allow the OR operator to also supply a default value to a `maybe` value by adding a new overload for the OR operator.

If it were a function `(|)`, then

```ts
assert type (|) : [t] maybe[t] -> t -> t
```

The second operand should be lazily evaluated. This should not cause an infinite loop:

```ts
let firstYes = [[t] maybes:list[maybe[t]]] -> maybe[t] {
  if let <yes item> = maybes[0] {
    return item | firstYes(maybes |> subsection(1, len(maybes)))
  } else {
    return none
  }
}

assert value firstYes([none, none, yes(2), yes(3)]) == yes(2)
assert value firstYes([none, none]) == none
```

Should work well in conjunction with #263

### LINKED ISSUE #266: Remove nonstandard `var`
Despite not being covered by the tests, both implementations for some reason support a "`var` statement," which does not exist in N. They should be removed.

https://discord.com/channels/710932856251351111/710936824121655396/883209709346254875

### LINKED ISSUE #269: Breaking Change: Fix `request` functions to all be in one function.
Currently, for all different types of requests, there are different functions, when ideally we can turn it into just one function.

### LINKED ISSUE #270: Breaking change: Change enum syntax
When Ash changed the function call syntax to C-style function calls, he forgot to do the same to enum variants in enum declarations and patterns. We should probably change that:

```ts
type maybe[t] =
	| none
	| yes(t)

type myType = new(str, int, float)

// Enum pattern can be used here because there's only one variant
let new(a, b, c) = new("hi", 3, 3.14)

if let yes(value) = yes("sir") {
  assert value value == "sir"
}
```

### LINKED ISSUE #272: Add `mut` modifier
Add `mut` modifier, similar to rust.

### LINKED ISSUE #273: Add dynamic imports
Since built-in modules already have a `_values` function we can easily make this work with other files that are not built in for compatability.

### LINKED ISSUE #274: Add an exit function
This function would allow an int to be passed in as an exit code and can be used to signify to other programs that the code has failed or has finished without error, this would also be useful for non-recoverable errors.

### LINKED ISSUE #276: Add Internal Traits
This will reduce the amount of inbuilt functions we have and make it more straightforward what is going on, this will cause all types to have constructors but this will be not very problematic as can be simple and straightforward, this will also allow us to have class methods that have the same name but different functions, something that we were unable to do before with the current way inbuilt functions are.

### LINKED ISSUE #277: Breaking Change: Remove unneeded internal functions
Depends on #276 

Once #276 is implemented we can delete old functions and only keep the ones that are needed, this will help users understand the code more clearly and will allow for us to have more useful inbuilt functions.

### LINKED ISSUE #278: Allow `match` to deconstruct enum values
This will help with enums as the only way to deconstruct one currently is `if let`, but with match it can be a lot cleaner:
```rs
match maybe_value {
    yes(val) -> print(val)
    none -> print("None")
}
```
instead of
```rs
if let yes(val) = maybe_value {
    print(val)
} else {
    print("None")
}

### LINKED ISSUE #279: Breaking Change: Remove `=` as a comparison and `not` as a keyword
This will create a lot of confusion once N released 

## `=` as a comparison
`=` is usually used for assigning values and it will be very confusing to the people using N if they see `=` being used as a comparison, and it might cause syntactical ambiguity somewhere down the line, this is not even very hard to replace as we already accept `==` as a comparison

## `not` as a keyword
If we are going to keep `not` as a keyword we should also add `and` and `or` for consistency, otherwise there is no reason for `not` to be in N at all given that we have `~`

Notifying @SheepTester on this.

### LINKED ISSUE #280: `match` requires input type to be hashable
This means that lists will cause a runtime error if they are put into the `match` statement, this is because match uses a dictionary to find the value to match to. I do not know a better way of doing this.

### LINKED ISSUE #281: If an incorrectly typed argument is passed into a function, it will throw an error.
There is a runtime error that occurs because the `token_or_tree` value passed in is not a token or tree but in fact a type.

### LINKED ISSUE #282: Breaking Change: Change Record syntax.
This is because it is confusing for new users, who have to deal with using newlines or `;` instead of `,`

### LINKED ISSUE #283: Breaking Change: Remove old `for` loop syntax
This will remove the old way of writing `for` loops, which was inefficient and was only left in for backwards compatibility

### PR DESCRIPTION
**Pull request by Ashvin-Ranjan**
closes #264
closes #257
closes #263
closes #259
closes #238
closes #239
closes #266
closes #260
closes #225
closes #218
closes #270 
closes #269 
closes #272 
closes #273
closes #274 
closes #279 
closes #280 
closes #278 
closes #281 
closes #276
closes #282
closes #277
closes #283

### CODE REVIEW

**Comment by Ashvin-Ranjan** [comment]
Awaiting on function tests to be updated

File: tests/syntax/function-expressions-and-better-patterns.n [hunk 1]
```diff
@@ -90,7 +90,7 @@ let funcs = (
 	// Although redundant, you can put as many parentheses around expressions,
 	// types, and patterns as you wish. With the new function expression syntax,
 	// you can also have parentheses around the name-type pair.
-	((((arg1)): ((type1))) -> (type2) { return (arg1) }),
+	(((arg1): (type1)) -> (type2) { return (arg1) }),
```
Comment (SheepTester) [inline tests/syntax/function-expressions-and-better-patterns.n:93]: Uh I think `((arg1))` should be kept. 
image

Comment (Ashvin-Ranjan) [reply tests/syntax/function-expressions-and-better-patterns.n:93]: I disagree

Comment (SheepTester) [reply tests/syntax/function-expressions-and-better-patterns.n:93]: Wouldn't that make the grammar more complex though

**Comment by anematode** [comment]
godly

**Pull request merged**